woocommerce/woocommerce-android
# Add helper functions to make convertion from EventBus events to coroutines easier

### Description
While working recently on the login parts, I found myself interacting with EventBus and our `Dispatcher` more than before, and I was using the following patter for converting the interaction to coroutine:
```kotlin
suspend fun callDispatcher() = suspendCancellableCoroutine { cont ->
   val listener = object {
       @Subscribe
       fun onSomeEvent(event: SomeEvent) {
           dispatcher.unregister(this)
           cont.resume(event)
        }
    }
   dispatcher.register(listener)
   dispatcher.dispatch(someAction)
   cont.invokeOnCancellation { dispatcher.unregister(listener) }
}
```

This works quite well, and IMO it's better than the previous pattern we were using, as it needed saving the `continuation` objects in the repository, and was resulting on some CancellationExceptions.
In this PR, I'm attempting to add some helper functions to make using this pattern even easier, it's in the form of two extensions, both at the file `EventBusCoroutines`.

On the next PR, I'll add an extension that converts events that fire multiple times to Flow, and I'll use it in `MediaFilesRepository` and `PluginRepository`.

### Testing instructions
Testing all the parts that changed would take a lot of time, but IMO, just testing some scenarios is enough to confirm that the conversion works, then reviewing the code well to make sure I didn't miss anything.

For a test scenario that covers both extensions, please follow these steps:

1. Use a WordPress.com that's not connected to any site.
2. Clear the app's data (or remove the concerned site from the DB manually) 
3. Open the app, then login using WordPress.com account.
4. Notice the "No Stores" error screen.
5. Click on "Enter site address"
6. Enter the site address URL, and then the site credentials.
7. Notice the Jetpack Connection screen.
8. Confirm the presence of WordPress.com consent at the bottom of the screen.
9. Click on "Connect Jetpack to your account"
10. Confirm that a WebView is opened.

The above goes through the function `AccountMismatchRepository#checkJetpackConnection` which uses both extensions.
